sdk/go/common/resource/asset.go

```diff
@@ -1108,9 +1110,9 @@ func (a *Archive) EnsureHash() error {
 				return err
 			}
 		} else {
-			// Otherwise, it's not an archive; we'll need to transform it into one.  Pick tar since it avoids
-			// any superfluous compression which doesn't actually help us in this situation.
-			err := a.Archive(TarArchive, hash)
+			// Otherwise, it's not an archive; we'll need to transform it into one.
+			// Pick ZIP as default for compat with lambda.
+			err := a.Archive(ZIPArchive, hash)
```

Comment:
I'm still confused why this is needed.

This code doesn't actually produce an archive, it just computes the hash.  Why does it matter whether we use TarArchive or ZipArchive for computing the hash?

We do ultimately provide a ZIP to Lambda per:

https://github.com/pulumi/pulumi-terraform-bridge/blob/25b33bb489e0075e7733a7ee5a4c6bf33a08b8e6/pkg/tfbridge/assets.go#L164-L164

Are you *sure* this change is necessary as part of the fix?

Reply:
@lukehoban I'm sure that if you remove this line, that we get the same error as without this change when running the lambda:
```sh
Evans-MBP:aws-go-lambda evanboyle$ aws lambda invoke --function-name $(pulumi stack output lambda) --region $(pulumi config get aws:region) --payload '"foo"' output.json
{
    "StatusCode": 200,
    "FunctionError": "Unhandled",
    "ExecutedVersion": "$LATEST"
}

Evans-MBP:aws-go-lambda evanboyle$ cat output.json 
{"errorMessage":"fork/exec /var/task/handler: permission denied","errorType":"PathError"}
```

With this fix:

```sh
Evans-MBP:aws-go-lambda evanboyle$ aws lambda invoke --function-name $(pulumi stack output lambda) --region $(pulumi config get aws:region) --payload '"foo"' output.json
{
    "StatusCode": 200,
    "ExecutedVersion": "$LATEST"
}
Evans-MBP:aws-go-lambda evanboyle$ cat output.json 
"FOO"
Evans-MBP:aws-go-lambda evanboyle$ 
```

Feel free to try it out locally. I'm not sure why this is necessary, but `EnsureHash` does store the resulting hash on the archive object. 

We do write out the hash when serializing as well: https://github.com/pulumi/pulumi/blob/84759ce01dbc1717a499a92dba924a780db12ce6/sdk/go/common/resource/asset.go#L227